### ISSUE TITLE
v2.0.0 - Clear cache example doesn't work

### ISSUE THREAD

**Issue by o-alexandrov** [vercel/swr#2281]
# Bug report

Clear cache example from the docs doesn't work

## Description / Observed Behavior

The cache still exists for the cache keys.
Ex. a user who logged out would see data from the previous session prior to revalidation. (on another page, so it's not about race conditions)

## Expected Behavior

To be equivalent to what worked in prior versions.

## Repro Steps / Code Example

I used to do in `swr@2.0.0-beta.6`:

```ts
import { useSWRConfig } from "swr"

const ReactComponent = () => {
  const { cache } = useSWRConfig()
  
  React.useEffect(() => {
    cache.clear()
  }, [cache]) // with some other dependencies
}
```

After updating to `2.0.0`, I try to achieve the same result using the code snippet [from the docs](https://swr.vercel.app/docs/mutation#mutate-multiple-items):

```ts
import { mutate } from "swr"

const clearCache = () => mutate(
  () => true,
  undefined,
  { revalidate: false }
)

// ...clear cache on logout
clearCache()
```
## Additional Context

SWR version. `2.0.0`

**Comment by huozhi** [comment]
You can checkout the partial mutate method to update the cache https://swr.vercel.app/docs/mutation#mutate-multiple-items

```js
const clearCache = () => mutate(
  () => true,
  undefined,
  { revalidate: false }
)

// ...clear cache on logout
clearCache()
```

**Comment by huozhi** [comment]
> The cache still exists for the cache keys.

Are you relying on cache keys to do sth with UI?

**Comment by o-alexandrov** [comment]
@huozhi in my manual testing, I made the following steps:
- upgrade from `2.0.0-beta.6` to `2.0.0` and make a code change mentioned in the description
- then, perform the following steps for the user:
  - log in as `USER_1`
  - open notifications page (swr loads notifications)
  - open profile page
  - log out (swr clears cache)
  - user is redirected to Homepage
  - log in as `USER_2`
  - open notifications page (swr briefly shows previous user's notifications because of the polluted cache)

Basically, I assume the issue happens to every user of `swr` who does what the docs say.
- the bad part here is the removal of `cache.clear` API, so upgrading from `2.0.0-beta.6` is impossible

---

Another issue with the new API using mutate is:
- there's no way to do `revalidate: false`, while triggering `revalidate` for cache keys visible on the current page
  - whereas `cache.clear()` resulted in revalidation on the current page

**Comment by o-alexandrov** [comment]
May I propose to add removed API back to `swr`?

```ts
const { cache } = useSWRConfig()
cache.clear()
```

**Comment by o-alexandrov** [comment]
After playing with it in [this repro (on CodeSandbox)](https://codesandbox.io/s/swr-bug-swr-swrinfinite-mutate-affects-different-cache-keys-khk221?file=/src/App.tsx), I realized:
- `mutate` from `swr`
- and `mutate` from `swr/infinite`

Mutate different keys, so `mutate` from `swr` won't affect any keys from `swr/infinite`.

**Comment by huozhi** [comment]
We skipped it by design since the infinite keys provided by users are function. We're now discussing the solution to fix it

**Comment by stack-pop** [comment]
I have also encountered this problem, may I ask when it can be fixed?

**Comment by sbbarragan** [comment]
Do you know if there's any progress on this discussion?

I'm using this way to clear all cache:
```js
const { cache } = useSWRConfig();

const clearCache = () => [...cache.keys()].forEach((key) => cache.delete(key));
```

**Comment by jansepke** [comment]
> Do you know if there's any progress on this discussion?
> 
> I'm using this way to clear all cache:
> 
> ```js
> const { cache } = useSWRConfig();
> 
> const clearCache = () => [...cache.keys()].forEach((key) => cache.delete(key));
> ```

but this will only delete the data from the cache, but not trigger a re-validation like `mutate()` is doing. still searching for a solution to invalidate and refetch the data of useInfinite using `useSWRConfig`

**Comment by jansepke** [comment]
Finally I found a solution:

```js
const { mutate, cache } = useSWRConfig();

const refresh = async () => {
    await Promise.all([...cache.keys()].map((key) => mutate(key)));
};
```

I found this hint in the source code: https://github.com/vercel/swr/blob/f0d80d32f7e6181b870b91fcaf5272084fbbbb45/_internal/src/utils/mutate.ts#L79

**Comment by o-alexandrov** [comment]
Closing, as the following works for me:

```ts
const { cache } = useSWRConfig();
for (const key of cache.keys()) cache.delete(key)
```

**Comment by jackirvine97** [comment]
> We skipped it by design since the infinite keys provided by users are function. We're now discussing the solution to fix it

Are there any updates on the official plan to resolve this issue?

Cache clearing ability is vital for maintaining data integrity in production apps with sign-in / sign-out functionality

**Comment by bodinsamuel** [comment]
hey team,
It's 2024 this issue is closed but it's not solved nor documented

**Comment by BooDzioo** [comment]
Hey team,
It's 2025 and this issue is closed but not solved nor documented :)

**Comment by zanderlan-n-ws** [comment]
Hey team,
It's 2025 and this issue is closed but not solved nor documented :)
